common/src/lib.rs

```diff
@@ -12,67 +12,61 @@ pub type Timestamp = u64;
     ///
     /// The order here is the leader selection priority order of the validators
-    /// which directly affects the result of `calculate_net_validator_set()`.
-    pub validator_set: Vec<(PublicKey, VotingPower)>,
-    /// The protocol version that must be used from next block.
+    /// which directly affects the effective validator set.
     ///
-    /// It must be a valid semantic version (e.g., `0.2.3`).
-    pub version: String,
-    /// The delegation state (delegator, delegatee) where indices to `validator_set` is given.
-    pub delegation: Vec<(usize, usize)>,
+    /// The `(usize, usize)` is `(first delegatee, current delegatee)`.
+    /// Two could differ if the first delegatee delagtes to other validators.
+    pub original_validator_set: Vec<(PublicKey, VotingPower, Option<(usize, usize)>)>,
+    // TODO: add various conditions for each delegation.
+    // - Unlock-Automatically-After-N-Blocks
+    // - Unlock-Automatically-After-T-Seconds
+    // - Unlock-If-The-Delegatee-Is-Not-Active
+    // - Unlock-If-The-Validator-Set-Changes
 }
 
-impl EssentialState {
-    /// Calculate the actual set of validators & voting power for the next block.
-    ///
-    /// The order here is same as the order of leaders in each round.
-    /// returns `None` if the delegation is not valid.
-    pub fn calculate_net_validator_set(&self) -> Result<Vec<(PublicKey, VotingPower)>, String> {
-        // check delegation by delegatee (which's forbidden)
-        let mut delegatee_indices = BTreeSet::new();
-        for (delegator_index, delegatee_index) in &self.delegation {
-            if delegatee_indices.contains(delegator_index) {
-                let delegatee: &PublicKey = &self.validator_set[*delegator_index].0;
-                return Err(format!("delegatee ({}) can't delegate", delegatee));
-            }
-            delegatee_indices.insert(*delegatee_index);
-        }
+impl DelegationState {
+    pub fn hash(&self) -> Hash256 {
+        Hash256::hash(serde_json::to_vec(self).unwrap())
+    }
 
-        // calculate the result
-        let mut validator_set: BTreeMap<_, _> = self.validator_set.iter().cloned().collect();
-        for (delegator_index, delegatee_index) in &self.delegation {
-            let delegator: &PublicKey = &self.validator_set[*delegator_index].0;
-            let delegator_voting_power = if let Some(x) = validator_set.get(delegator) {
-                *x
-            } else {
-                return Err(format!("delegator not found: {}", delegator));
-            };
-            let delegatee: &PublicKey = &self.validator_set[*delegatee_index].0;
-            let delegatee_voting_power = if let Some(x) = validator_set.get(delegatee) {
-                *x
+    pub fn calculate_effective_validator_set(
+        &self,
+    ) -> Result<Vec<(PublicKey, VotingPower)>, String> {
+        let mut validator_set = BTreeMap::new();
+        for (public_key, voting_power, delegation) in &self.original_validator_set {
```

Comment:
```suggestion
        for (public_key, voting_power, delegation_status) in &self.original_validator_set {
```

Reply:
Can you explain the reason why and when?

Reply:
@TomTaehoonKim 
https://github.com/postech-dao/simperby/pull/60#discussion_r961738103

Reply:
@junha1 understood.